### PR TITLE
Craft request uri object

### PR DESCRIPTION
**Pull request by JanDeDobbeleer**
This has been deployed on our production test environment: demo.failforward.tech.
The Request URI payload comes from the following function in the lambda:

```golang
func getRequestData() string {
	itsmeClient := itsme.NewBasicItsmeClient(&environment{})
	config := itsme.RequestURIConfiguration{
		URLConfiguration: itsme.URLConfiguration{
			RedirectURI: "https://backend.failforward.tech/production/redirect",
			Scopes:      []string{"email", "picture", "profile", "address"},
			ServiceCode: "BMIDITP_LOGIN",
		},
		AcrValue: itsme.ACRAdvanced,
		Nonce:    getRandomString(50),
		State:    "thisismyriflethisismygun",
		Claims:   []itsme.Claim{itsme.ClaimCityOfBirth, itsme.ClaimEid, itsme.ClaimDevice, itsme.ClaimNationality, itsme.ClaimPhoto},
	}
	jwt, err := itsmeClient.CreateRequestURIPayload(config)
	if err != nil {
		return err.Error()
	}
	return jwt
}
```

I'm open to suggestion on how to properly test this. One idea is to extract the object -> JSON creation and validate that. A second one might be to create a test spec so we can encrypt these values in a test environment (although that logic isn't exactly ours, so not sure about that).

One thing I'm not certain about is if we have to expose the `Nonce`.

### CODE REVIEW

File: oidc.go [hunk 1]
```diff
@@ -132,7 +174,32 @@ func (oidc *openIDClient) createRequestURIPayload(payload string) (string, error
...
+		Iss:          clientID,
+		Nonce:        config.Nonce,
+		State:        config.State,
+		Claims: userInfo{
+			Userinfo: make(map[Claim]interface{}),
+		},
+	}
+	for _, claim := range config.Claims {
+		requestURIData.Claims.Userinfo[claim] = nil
+	}
```
Comment (RoelandMatthijssens) [inline oidc.go]: - Extract the method `addClaimsToRequestURI(RequestURI, claims)` so we can test it in unit.
- Don't add the claims to the requestURI object here but extract it to the new function as well
- in the extracted method check if we have claims and then add them to the requestURI

File: oidc.go [hunk 2]
```diff
@@ -132,7 +174,32 @@ func (oidc *openIDClient) createRequestURIPayload(payload string) (string, error
...
+	data, err := json.Marshal(requestURIData)
+	if err != nil {
+		return "", errors.New("Unable to marchal request URI data")
+	}
+	return oidc.encrypter.getSerializedSignedEncryptedJWT(signingKey, ITSMEEncryptionKey, data)
```
Comment (RoelandMatthijssens) [inline oidc.go]: Don't return a signed string, return it unsigned so we can test this method. Then in the calling function take care of the signing.

File: oidc.go (hunk 1 above)
Comment (JanDeDobbeleer) [reply oidc.go]: Not sure what the added benefit is of that? It's a simple map which needs to be validated by what it does in terms of JSON parsing (that's what it's actually for).

File: oidc_test.go [hunk 3]
```diff
@@ -119,3 +125,88 @@ func TestGetOIDCAuthenticationUrlWithUrlEncoding(t *testing.T) {
...
+
+type args struct {
+	config   *RequestURIConfiguration
+	aud      string
+	scope    string
+	clientID string
+}
+
+func createJSONByteResponse(args *args, claims []Claim) []byte {
+	template := "{\"acr_values\":\"%s\",\"aud\":\"%s\",\"claims\":{\"userinfo\":{%s}},\"client_id\":\"%s\",\"iss\":\"%s\",\"nonce\":\"%s\",\"redirect_uri\":\"%s\",\"response_type\":\"code\",\"scope\":\"%s\",\"state\":\"%s\"}"
```
Comment (RoelandMatthijssens) [inline oidc_test.go]: Apparently you can use backticks to start multiline strings in go. so you don't have to escape every `"`

Comment (JanDeDobbeleer) [reply oidc_test.go]: True, but if I'm not mistaken that also adds spaces in which case the tests will break as well.

**Pull request merged**